### PR TITLE
[FAB-17993] Add Support For CouchDB 3.x

### PR DESCRIPTION
**Pull request by lindluni**
Couch 3.x requires the user provide an admin user at startup. It also prevents the DB user from modifying the `_users` security objects. This affects the peer startup behavior where it checks if the `_users` table exists and creates the table. Since the user no longer has permission to access the `_users` security object a check was added to confirm the DB version before attempting to modify the `_users` table.

Tests were updated to included an admin user and password for couchDB in the DB configs as well as on the CouchDB runner

Signed-off-by: Brett Logan <brett.t.logan@ibm.com>

### CODE REVIEW

File: Makefile
Comment (cendhu) [inline Makefile]: As per the CouchDB doc (which we may not trust 😄 based on what we found out), [the upgrade from 2.3 to 3.1](https://docs.couchdb.org/en/stable/install/upgrading.html#standalone-single-node-upgrades) looks very similar to how the peers are being upgraded.  Another option is to simply start a new CouchDB instance and the peer will build the DB automatically.

Is there a reason for supporting both 2.3 & 3.1? something related to release version or handling risk or avoid statedb rebuild?

For CI and unit-test/integration test, should we need to repeat the run on both CouchDB 2.3 and CouchDB 3.1 to ensure we aren't breaking any backward compatibility in the ledger code?

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb_test_export.go
Comment (cendhu) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb_test_export.go]: As it is present in the `couchdb_test_export.go`, I have to assume that the test would be run only using CouchDB 2.3. I don't know what problem we would face/introduce due to not testing with both CouchDB 2.3 and 3.1.

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go
Comment (cendhu) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go]: It might be possible to modify _users if appropriate config is set -- as per the depreciation notice.

I am fine with the current code as it helps to remove having an assumption on the config.

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go
Comment (cendhu) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go]: Do we need hash-graph's version matching pkg? I thought we would always support one CouchDB version. At the worst case, we might support the previous and current version. Can't we simply compare strings (`"3.1.0", "2.3"`) rather than using a new dependency and making this code generic? Further, LessThan might return true even if the deployment uses `2.0` CouchDB. Hence, I am confused a bit. If there is a plan to support multiple CouchDB versions (more than 2) officially, it would create complication for us whenever we touch `statecouchdb` pkg.

**Comment by lindluni** [comment]
@manish-sethi @cendhu 

I added a unit-test to confirm legacy `_users` functionality still works using a 2.3  couchdb instance.

Do you think I should add a legacy 2.3 E2E integration test? I can run a quick E2E smoke test across a network with couchdb 2.3 just to confirm functionality. Or I can add a second peer to one of the existing integration tests that uses a 2.3 couch instance for its StateDB.

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go
Comment (cendhu) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go]: Shouldn't it be `3.1.0`?
Screenshot 2020-06-17 at 9 04 34 AM

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go
Comment (cendhu) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go]: Not related to changes made in this PR. We need to understand the last bullet in https://docs.couchdb.org/en/latest/whatsnew/3.0.html#performance better and first bullet in https://docs.couchdb.org/en/latest/whatsnew/3.0.html#deprecated-feature-removal.

**Review by cendhu — COMMENTED** [review]
Thanks, Brett. I have few doubts which I have raised in the PR. Once I understand the scope of this PR, I will do an in-depth review.

**Comment by cendhu** [comment]
> Do you think I should add a legacy 2.3 E2E integration test? I can run a quick E2E smoke test across a network with couchdb 2.3 just to confirm functionality. Or I can add a second peer to one of the existing integration tests that uses a 2.3 couch instance for its StateDB.

I was reviewing the PR and saw this after submitting my reviews. I had raised a similar question in one my comment. 

We do use CouchDB at the `core/ledger/kvledger/tests` pkg which is a ledger level integration test for us to ensure backward compatibility since v1.0. In every major release, we add complete ledgerData folder to ensure we aren't breaking anything. If we plan to support two versions of CouchDB, it may be necessary to run these test on both (brute force approach) or the CouchDB version may not be matter (if we think carefully and analyse the complete impact).

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go
Comment (lindluni) [reply core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go]: With this we would support 3.1.x and 2.3.x. Today in fabric we actually test against 2.3 (though there is no actual 2.3, its whatever the latest 2.3.x version is), so it's whatever is latest, no idea until runtime what the patch level is. And I actually only chose to switch to the semver package because it seemed less error prone than doing string splits and comparisons. I also only elected to use it as the package is already part of fabrics dep tree.

If there is other dissent on this I'll switch it to a string matcher

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go
Comment (lindluni) [reply core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go]: I'll clarify the comment, and point to the deprecation notice.

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb_test_export.go
Comment (lindluni) [reply core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb_test_export.go]: The proposal here is all testing is done using 3.1, the only testing done on 2.3 is the auto-creation of the `_users` table. Unless we think it should be reversed, all testing on 2.3, with only the _users case tested on 3.1. But my opinion is that we should move forward and 3.1 should become our new standard.

File: Makefile
Comment (lindluni) [reply Makefile]: We could drop support for 2.3 altogether in the 2.{2,next} stream, and then a lot of these changes go away. TBH, I would prefer that in 2.{2,next} we say the minimum required level of Couch is 3.1. But Couch 2 has been around for nearly 10 years (so getting users to migrate may not be easy, as 3.x is not as tested), and our code is not backwards compatible if we remove support for 2.3, so that would require a major version change to rip out the creation of the `_users` table (which is required for removing support for 2.3).

**Comment by lindluni** [comment]
> > Do you think I should add a legacy 2.3 E2E integration test? I can run a quick E2E smoke test across a network with couchdb 2.3 just to confirm functionality. Or I can add a second peer to one of the existing integration tests that uses a 2.3 couch instance for its StateDB.
> 
> I was reviewing the PR and saw this after submitting my reviews. I had raised a similar question in one my comment.
> 
> We do use CouchDB at the `core/ledger/kvledger/tests` pkg which is a ledger level integration test for us to ensure backward compatibility since v1.0. In every major release, we add complete ledgerData folder to ensure we aren't breaking anything. If we plan to support two versions of CouchDB, it may be necessary to run these test on both (brute force approach) or the CouchDB version may not be matter (if we think carefully and analyse the complete impact).

There are quite a few integration scenarios I can think through here. We will need to have a discussion on a plan at large for what we want to do, wrt integration tests and version support

File: Makefile
Comment (ale-linux) [reply Makefile]: I'm not a ledger expert so I'll defer to @cendhu, @manish-sethi and @denyeart. However, my recommendation would be
- to only upgrade couch versions with a good reason (e.g. a new feature of the new version of couch is necessary; or the old couch version will no longer receive security updates, or it has reached EoL)
- to only upgrade couch version across (major) releases
- to never support more than one couch version per release

With this in mind, I would recommend (once again, deferring to the expert) not to go into the business of supporting multiple versions at the same time.

Comment (cendhu) [reply Makefile]: @ale-linux Agree. There is a bug (http://danluu.com/fsyncgate/) (https://github.com/apache/couchdb/commit/3505281559513e2922484ebf0996a8846dcc0a34) that is fixed only in CouchDB 3.x and we do need them. At the same time, we are not sure about the stability of CouchDB 3.x (new release). We haven't decided yet on supporting multiple CouchDB versions (mostly, we shouldn't).

**Comment by denyeart** [comment]
I don't think we need all the complexity of separate v2.x and v3.x support.
I think we can keep the peer code generic between couchdb 2.x and 3.x. In 3.x admin can still create _users table, I think we just need to suppress the update on _users table that creates the security object. This can be suppressed on both 2.x and 3.x deployments. Basically just add one if statement to skip for _users table in https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go#L1426

**Comment by lindluni** [comment]
> I don't think we need all the complexity of separate v2.x and v3.x support.
> I think we can keep the peer code generic between couchdb 2.x and 3.x. In 3.x admin can still create _users table, I think we just need to suppress the update on _users table that creates the security object. This can be suppressed on both 2.x and 3.x deployments. Basically just add one if statement to skip for _users table in https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go#L1426

I refactored `createDatabaseIfNotExist()` to have a single entrypoint into `applyDatabasePermissions()` (as opposed to the existing three entrypoints) so I could perform the check there, rather than doing a check and no-op in  `createDatabaseIfNotExist()`

**Comment by lindluni** [comment]
I will add an integration test yet that does an e2e smoke test against 2.3

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go
Comment (denyeart) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go]: I think this debug is misplaced... it will print this message even if the database gets created in this function.

**Comment by lindluni** [comment]
Removed all references to the legacy 2.3 Couch. So this now fully supports Couch 3.1 and does all testing on 3.1

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go
Comment (manish-sethi) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go]: Since you inverted the condition, I find that explicitly checking `dbInfo == nil && couchDBReturn.StatusCode == 404` makes it more readable.

Just to take a note, the both parts of the condition are actually same but it seems that they were always like that.

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go
Comment (manish-sethi) [inline core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go]: You reversed the logic of "suppressing the error" into returning the error. However, the comment is outdated now...

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go
Comment (lindluni) [reply core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go]: Done

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go
Comment (lindluni) [reply core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go]: removed the comment

**Comment by lindluni** [comment]
opened #1430 to remove the 3.1 migration and only skip the security object modification. Review should move there. This change will automatically rebase itself on master once that change is merged (GitHub now autorebases when commits are merged from other PR's)

**Comment by lindluni** [comment]
@Mergifyio rebase

**Review by denyeart — APPROVED** [review]
I like this much better than the dual support of 2.3 and 3.1 PR :-)

**Pull request merged**
